set.mm

```diff
@@ -477442,6 +477720,87 @@ The sumset operation can be used for both group (additive) operations and
     ) ) $.
 
+  ${
+    $d R r $.
+    $( Value of the spectrum of the ring ` R ` .  (Contributed by Thierry
+       Arnoux, 2-Jun-2024.) $)
+    rspecval $p |- ( R e. Ring
+                 -> ( Spec ` R ) = ( ( IDLsrg ` R ) |`s ( PrmIdeal ` R ) ) ) $=
+      ( vr cv cidlsrg cfv cprmidl cress co crg wceq fveq2 oveq12d df-rspec ovex
+      crspec fvmpt ) BABCZDEZQFEZGHADEZAFEZGHIOQAJRTSUAGQADKQAFKLBMTUAGNP $.
+  $}
+
+  ${
+    rspecbas.1 $e |- S = ( Spec ` R ) $.
+    $( The prime ideals form the base of the spectrum of a ring.  (Contributed
+       by Thierry Arnoux, 2-Jun-2024.) $)
+    rspecbas $p |- ( R e. Ring -> ( PrmIdeal ` R ) = ( Base ` S ) ) $=
+      ( crg wcel cprmidl cfv cidlsrg cress cbs wceq clidl prmidlssidl idlsrgbas
+      co wss eqid sseqtrd ressbas2 syl crspec rspecval syl5eq fveq2d eqtr4d ) A
+      DEZAFGZAHGZUGIOZJGZBJGUFUGUHJGZPUGUJKUFUGALGZUKAMAUHULDUHQULQNRUGUKUIUHUI
+      QUKQSTUFBUIJUFBAUAGUICAUBUCUDUE $.
+
+    ${
+      $d I i $.  $d P j $.  $d R i j $.  $d V i $.
+      rspectset.1 $e |- I = ( LIdeal ` R ) $.
+      rspectset.2 $e |- P = ( PrmIdeal ` R ) $.
+      rspectset.3 $e |- J = ran ( i e. I |-> { j e. P | -. j C_ i } ) $.
+      $( Topology component of the spectrum of a ring.  (Contributed by Thierry
+         Arnoux, 2-Jun-2024.) $)
+      rspectset $p |- ( R e. Ring -> J = ( TopSet ` S ) ) $=
+        ( crg wcel cidlsrg cfv cts cprmidl cress cvv eqid co wceq fvex resstset
+        ax-mp idlsrgtset crspec rspecval syl5eq fveq2d 3eqtr4a ) BLMZBNOZPOZUMB
+        QOZRUAZPOZGCPOUOSMUNUQUBBQUCUOUMUPUNSUPTUNTUDUEABUMDEFGLUMTIJKUFULCUPPU
+        LCBUGOUPHBUHUIUJUK $.
+    $}
+  $}
+
+
+$(
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+  Prime Elements
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+$)
+
+
+$(
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+  Unique factorization domains
+-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-.-
+$)
+
+  $( Token for the class of unique factorization domains. $)
+  $c UFD $.
+
+  $( Class of unique factorization domains. $)
+  cufd $a class UFD $.
+
+  ${
+    $d i r $.
+    $( Define the class of unique factorization domains.  A unique
+       factorization domain (UFD for short), is a commutative ring with an
+       absolute value (from ~ abvtriv this is equivalent to being a domain)
+       such that every prime ideal contains a prime element (this is a
+       characterization due to Irving Kaplansky).  UFD are also sometimes also
+       called a "factorial rings" following the terminology of Bourbaki.
+       (Contributed by Mario Carneiro, 17-Feb-2015.) $)
+    df-ufd $a |- UFD = { r e. CRing | ( ( AbsVal ` r ) =/= (/) /\
+               A. i e. ( PrmIdeal ` r ) ( i i^i ( RPrime ` r ) ) =/= (/) ) } $.
+  $}
+
+  ${
+    $d A r $.  $d R i r $.
+    isufd.1 $e |- A = ( AbsVal ` R ) $.
```

Comment:
Why is a class variable provided for `( AbsVal `` R )` and not for `( PrmIdeal `` R )`and `( RPrime `` R )`? Since ` A` occurs only once, I think a class variable is not necessary here.

Reply:
You're right, this is not coherent.

Do we have any convention about when to use class variables or not?
I tend to try to use a class variable for every structure extractor, so here I might instead add one for both ``( PrmIdeal ` R )`` and ``( RPrime ` R )``!

Reply:
Ok so in my last commit, I ended up assigning a class variable to all 3:``( AbsVal ` R )``, ``( PrmIdeal ` R )`` and ``( RPrime ` R )``. This should allow the maximum flexibility (you can always use `eqid` that's not needed).

I hope it's fine too.